Adding upper momentum cut in FilterStepPointMomentum to be used in co…

oksuzian opened the pull request:
Addition of the upper momentum cut in FilterStepPointMomentum to be used in cosmic resampling jobs.
Addition of new variables in Analyses/src/StepPointMC1stHitDumper_module.cc
Other minor changes.

oksuzian commented on Analyses/src/StepPointMC1stHitDumper_module.cc:
```diff
@@ -195,7 +215,7 @@ namespace mu2e {
       // }
       if (stepPoints.size()) {
-        hit_ = VDHit(toff_, stepPoints.at(0));
+        hit_ = VDHit(toff_, event, stepPoints.at(0));
```
Hi Dave,

The StepPointMC1stHitDumper module is just an analyzer module that makes a simple ntuple from the first step point MC. I found it handy to make simple plots for diagnostics purposes.
The upper momentum cutoff is done in the filter module: FilterStepPointMomentum.
The upper momentum cut of 0.5 GeV removes all the straight tracks and it reduces the size of the files by about a factor 10. It will be useful to have this functionality for a large scale cosmics MC production.

I've submitted these changes in one commit, and they are completely unrelated.
Sorry, if it adds confusion.
Would it help to separate commits in the future?

brownd1978 commented:
Hi Yuri,
this is fine, thanks for the explanation.  I also had to put upper-momentum
cuts on the selections for mixed cosmics in MDC2018 for the same reason.
cheers, Dave

On Tue, Nov 5, 2019 at 7:48 AM oksuzian <notifications@github.com> wrote:

> *@oksuzian* commented on this pull request.
> ------------------------------
>
> In Analyses/src/StepPointMC1stHitDumper_module.cc
> <https://github.com/Mu2e/Offline/pull/46#discussion_r342616988>:
>
> > @@ -195,7 +215,7 @@ namespace mu2e {
>          // std::cout << event.id() << ": " << stepPoints.at(i) << std::endl;
>        // }
>        if (stepPoints.size()) {
> -        hit_ = VDHit(toff_, stepPoints.at(0));
> +        hit_ = VDHit(toff_, event, stepPoints.at(0));
>
> Hi Dave,
>
> The StepPointMC1stHitDumper module is just an analyzer module that makes a
> simple ntuple from the first step point MC. I found it handy to make simple
> plots for diagnostics purposes.
> The upper momentum cutoff is done in the filter module:
> FilterStepPointMomentum.
> The upper momentum cut of 0.5 GeV removes all the straight tracks and it
> reduces the size of the files by about a factor 10. It will be useful to
> have this functionality for a large scale cosmics MC production.
>
> I've submitted these changes in one commit, and they are completely
> unrelated.
> Sorry, if it adds confusion.
> Would it help to separate commits in the future?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/46?email_source=notifications&email_token=ABAH573ZKESC6EM6HPIKRULQSGITBA5CNFSM4JI5IWR2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCKK37OI#discussion_r342616988>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH576YQ4CN4KAG5QTSZ2DQSGITBANCNFSM4JI5IWRQ>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

gaponenko commented:
> Hi Yuri, can you please give a use case example for upper momentum cut?

I found this explanation in another post to this thread:

    The upper momentum cut of 0.5 GeV removes all the straight tracks and it reduces the size of the files by about a factor 

A particle above 0.5 GeV can create a CE-like delta ray, how is this handled with the cut?  Does not seem to be safe for physics.

oksuzian commented:
Hi Andrei,

The use case is described in:
JobConfig/cosmic/dsstops_resampler.fcl

The configuration in the fcl file above only saves events that have a track below 0.5 GeV in the tracker volume.
As Dave pointed out, it's similar to the momentum cut we already use 'physics.filters.DigiFilter.MaxParticleMom'. There we actually cut at 200 MeV:
JobConfig/mixing/CRY-cosmicMix.fcl:physics.filters.DigiFilter.MaxParticleMom : 200.0
JobConfig/mixing/DS-cosmicMix.fcl:physics.filters.DigiFilter.MaxParticleMom : 200.0

brownd1978 commented:
Just to make sure, do you look at all the StepPoints or just the first?
the first could be above threshold, then the particle interacts and a later
one is below threshold.

On Tue, Nov 5, 2019 at 9:00 AM oksuzian <notifications@github.com> wrote:

> Hi Andrei,
>
> The use case is described in:
> JobConfig/cosmic/dsstops_resampler.fcl
>
> The configuration in the fcl file above only saves events that have a
> track below 0.5 GeV in the tracker volume.
> As Dave pointed out, it's similar to the momentum cut we already use
> 'physics.filters.DigiFilter.MaxParticleMom'. There we actually cut at 200
> MeV:
> JobConfig/mixing/CRY-cosmicMix.fcl:physics.filters.DigiFilter.MaxParticleMom
> : 200.0
> JobConfig/mixing/DS-cosmicMix.fcl:physics.filters.DigiFilter.MaxParticleMom
> : 200.0
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/46?email_source=notifications&email_token=ABAH575V6I6U6IKEGXREQFDQSGRBFA5CNFSM4JI5IWR2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEDDOMHY#issuecomment-549905951>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH573A7TCG2HA3NMMTPO3QSGRBFANCNFSM4JI5IWRQ>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

oksuzian commented:
> Just to make sure, do you look at all the StepPoints or just the first? the first could be above threshold, then the particle interacts and a later one is below threshold.
> […](#)

Dave,

The FilterStepPointMomentum module Andrei originally wrote looks at all the step points, not the very first one.
Please see below.

      auto ih = event.getValidHandle<StepPointMCCollection>(cn);
      for(const auto& hit : *ih) {
        if(hit.momentum().mag() > cutMomentumMin_ && hit.momentum().mag() < cutMomentumMax_) {
          passed = true;
          break;
        }
      }

oksuzian commented:
Andrei,

I've updated 

> Hi Yuri, we are migrating Offline to validated fhicl. As Rob announced in a Software meeting some time ago, newly merged code should use the Config pattern rather than ParameterSet. Please update the filter module to use validation. There is some info here
> https://mu2ewiki.fnal.gov/wiki/FclIntro#The_C.2B.2B_API_.28Validated_FHiCL.29
> More at https://cdcvs.fnal.gov/redmine/projects/art/wiki/Configuration_validation_and_description
> and https://cdcvs.fnal.gov/redmine/projects/fhicl-cpp/wiki/Configuration_validation_and_fhiclcpp_types
> we also have multiple examples of Offline modules that are already migrated.

I've changed the code to use Config pattern per your request.
